### PR TITLE
raft: port transfer leader feature from etcd

### PR DESCRIPTION
**Pull request by tiancaiamao**
PTAL @siddontang  @BusyJay @ngaut 
and https://github.com/pingcap/kvproto/pull/53 too

### CODE REVIEW

**Comment by siddontang** [comment]
Add tag in log.

File: src/raft/raft.rs [hunk 1]
```diff
@@ -725,6 +742,57 @@ impl<T: Storage> Raft<T> {
...
+                return;
+            }
+            self.abort_leader_transfer();
+            info!("{} [term {}] abort transfer leadership to {}",
+                  self.id,
+                  self.term,
+                  last_lead_transferee);
+        }
+        if lead_transferee == self.id {
+            if self.lead_transferee.is_none() {
```
Comment (siddontang) [inline src/raft/raft.rs]: etcd uses last_lead_transferee to check here, is it ok if we use lead_transferee here?

File: tests/test_raft.rs [hunk 2]
```diff
@@ -1923,3 +1931,291 @@ fn test_commit_after_remove_node() {
...
 }
+
+// test_leader_transfer_to_uptodate_node verifies transferring should succeed
+// if the transferee has the most up-to-date log entires when transfer starts.
+#[test]
+fn test_leader_transfer_to_uptodate_node() {
+    let mut nt = Network::new(vec![None, None, None]);
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    let lead_id = nt.peers[&1].leader_id;
```
Comment (siddontang) [inline tests/test_raft.rs]: can we use a lead var for this test like `let lead = nt.peers.get(&1).unwrap()`?

**Comment by siddontang** [comment]
👍 

Rest LGTM

File: tests/test_raft.rs (hunk 2 above)
Comment (tiancaiamao) [inline tests/test_raft.rs]: no...in rust... we can't.
if lead is borrowed here, later when we call check_leader_transfer_state, compiler would
complain can't borrow mutable more than once.

File: src/raft/raft.rs (hunk 1 above)
Comment (tiancaiamao) [inline src/raft/raft.rs]: we did use a last_lead_transferee, several lines before here.
a little difference is etcd use lead_transferee uint64, equal to 0 means None. 
while I think use Option<u64> conform to rust Idioms.

Comment (tiancaiamao) [inline src/raft/raft.rs]: we did use a last_lead_transferee, several lines before here.
a little difference is etcd use lead_transferee uint64, equal to 0 means None. 
while I think use Option<u64> conform to rust Idioms.

Comment (siddontang) [inline src/raft/raft.rs]: https://github.com/coreos/etcd/pull/4916/files#diff-a8629a0eb6da1c6c3805111f1f1af85eR726

Comment (siddontang) [inline src/raft/raft.rs]: etcd uses `if lastLeadTransferee == None {` here.

File: src/raft/raft.rs [hunk 3]
```diff
@@ -668,6 +679,13 @@ impl<T: Storage> Raft<T> {
             }
             return Ok(());
         }
+        if m.get_msg_type() == MessageType::MsgTransferLeader && self.state != StateRole::Leader {
+            debug!("{} [term {} state {:?}] ignoring MsgTransferLeader to {}",
```
Comment (ngaut) [inline src/raft/raft.rs]: Please add tag in log.

File: src/raft/raft.rs [hunk 4]
```diff
@@ -885,6 +975,15 @@ impl<T: Storage> Raft<T> {
                     // drop any new proposals.
                     return;
                 }
+                if self.lead_transferee.is_some() {
+                    debug!("{} [term {}] transfer leadership to {} is in progress; dropping \
```
Comment (ngaut) [inline src/raft/raft.rs]: Ditto.

File: tests/test_raft.rs [hunk 5]
```diff
@@ -1923,3 +1931,291 @@ fn test_commit_after_remove_node() {
...
+               lead_transferee,
+               3);
+    }
+
+    nt.peers.get_mut(&1).unwrap().remove_node(3);
+
+    check_leader_transfer_state(nt.peers.get(&1).unwrap(), StateRole::Leader, 1);
+}
+
+// TestLeaderTransferBack verifies leadership can transfer
```
Comment (ngaut) [inline tests/test_raft.rs]: s/TestLeaderTransferBack/test_leader_transfer_back/

File: tests/test_raft.rs [hunk 6]
```diff
@@ -1923,3 +1931,291 @@ fn test_commit_after_remove_node() {
...
+               3);
+    }
+
+    // Transfer leadership back to self.
+    nt.send(vec![new_message(1, 1, MessageType::MsgTransferLeader, 0)]);
+
+    check_leader_transfer_state(nt.peers.get(&1).unwrap(), StateRole::Leader, 1);
+}
+
+// TestLeaderTransferSecondTransferToAnotherNode verifies leader can transfer to another node
```
Comment (ngaut) [inline tests/test_raft.rs]: s/TestLeaderTransferSecondTransferToAnotherNode/test_leader_transfer_second_transfer_to_another_node/

File: tests/test_raft.rs [hunk 7]
```diff
@@ -1923,3 +1931,291 @@ fn test_commit_after_remove_node() {
...
+               3);
+    }
+
+    // Transfer leadership to another node.
+    nt.send(vec![new_message(2, 1, MessageType::MsgTransferLeader, 0)]);
+
+    check_leader_transfer_state(nt.peers.get(&1).unwrap(), StateRole::Follower, 2);
+}
+
+// TestLeaderTransferSecondTransferToSameNode verifies second transfer leader request
```
Comment (ngaut) [inline tests/test_raft.rs]: s/TestLeaderTransferSecondTransferToSameNode/test_leader_transfer_second_transfer_to_same_node/

File: src/raft/raft.rs [hunk 8]
```diff
@@ -159,6 +159,10 @@ pub struct Raft<T: Storage> {
     /// the leader id
     pub leader_id: u64,
 
+    // lead_transferee is id of the leader transfer target when its value is not None.
```
Comment (BusyJay) [inline src/raft/raft.rs]: We should use doc comment here.

File: src/raft/raft.rs [hunk 9]
```diff
@@ -793,7 +870,18 @@ impl<T: Storage> Raft<T> {
         }
         match m.get_msg_type() {
             MessageType::MsgAppendResponse => {
-                self.handle_append_response(m, old_paused, send_append, maybe_commit)
+                self.handle_append_response(m, old_paused, send_append, maybe_commit);
+                // Transfer leadership is in progress.
+                if let Some(lead_transferee) = self.lead_transferee {
```
Comment (BusyJay) [inline src/raft/raft.rs]: This block should be moved to L758.

File: src/raft/raft.rs [hunk 10]
```diff
@@ -1185,6 +1304,12 @@ impl<T: Storage> Raft<T> {
         if self.maybe_commit() {
             self.bcast_append();
         }
+        // If the removed node is the lead_transferee, then abort the leadership transferring.
+        if let Some(lead_transferee) = self.lead_transferee {
```
Comment (BusyJay) [inline src/raft/raft.rs]: Why not just use `self.lead_transferee == Some(id)`?

File: src/raft/raft.rs [hunk 11]
```diff
@@ -1251,4 +1376,15 @@ impl<T: Storage> Raft<T> {
         }
         act >= self.quorum()
     }
+
+    pub fn send_timeout_now(&mut self, to: u64) {
+        let mut msg = Message::new();
```
Comment (BusyJay) [inline src/raft/raft.rs]: Use `new_message` instead.

File: tests/test_raft.rs [hunk 12]
```diff
@@ -1923,3 +1931,291 @@ fn test_commit_after_remove_node() {
...
+
+// test_leader_transfer_to_uptodate_node verifies transferring should succeed
+// if the transferee has the most up-to-date log entires when transfer starts.
+#[test]
+fn test_leader_transfer_to_uptodate_node() {
+    let mut nt = Network::new(vec![None, None, None]);
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    let lead_id = nt.peers[&1].leader_id;
+    if lead_id != 1 {
```
Comment (BusyJay) [inline tests/test_raft.rs]: Just use `assert_eq` is ok.

**Comment by xiang90** [comment]
If you have made anything different from etcd/raft, can you send us a pr? or can you give me a summary so i can port the changes back to etcd/raft if necessary.

**Comment by ngaut** [comment]
Sure, @xiang90 
We haven't change the implementation of raft.

**Comment by tiancaiamao** [comment]
PTA @siddontang @ngaut @BusyJay

File: tests/test_raft.rs [hunk 13]
```diff
@@ -1923,3 +1931,289 @@ fn test_commit_after_remove_node() {
...
+    cs.set_nodes(nt.peers[&1].nodes());
+    nt.storage[&1]
+        .wl()
+        .create_snapshot(nt.peers[&1].raft_log.applied, Some(cs), vec![])
+        .expect("");
+    nt.storage[&1].wl().compact(nt.peers[&1].raft_log.applied).expect("");
+
+    nt.recover();
+    let matched = nt.peers[&1].prs[&3].matched;
+    if matched != 1 {
```
Comment (BusyJay) [inline tests/test_raft.rs]: Just use `assert_eq` instead.

**Comment by ngaut** [comment]
LGTM.
PTAL @BusyJay

File: tests/test_raft.rs [hunk 14]
```diff
@@ -1923,3 +1931,279 @@ fn test_commit_after_remove_node() {
...
+fn test_leader_transfer_timeout() {
+    let mut nt = Network::new(vec![None, None, None]);
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+
+    nt.isolate(3);
+
+    // Transfer leadership to isolated node, wait for timeout.
+    nt.send(vec![new_message(3, 1, MessageType::MsgTransferLeader, 0)]);
+    let lead_transferee = nt.peers[&1].lead_transferee.unwrap();
+    if lead_transferee != 3 {
```
Comment (BusyJay) [inline tests/test_raft.rs]: `assert_eq`

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**